src/gtc/oir.py

```diff
@@ -159,7 +159,7 @@ def full(cls):
         return cls(start=AxisBound.start(), end=AxisBound.end())
 
 
-class HorizontalExecution(LocNode):
```

Comment:
Why does this fix #559?

Reply:
The problem is currently that the horizontal execution merging when applied in a nested way results in multiple `HorizontalExecutions` having the same `oir.LocalScalar` (a decl). Therefore it ends up with multiple declarations of the same symbol name. The symbol table collects for the sub-tree, skipping iterating into new symbol tables. Therefore making `HorizontalExecution` a `SymbolTableTrait` removes the duplicate decls.

Reply:
I agree that it makes sense to open a scope at the HorizontalExecution level, however it feels like it is solving the issue by side-effect of how the merging is implemented. I think merging should respect that there are 2 different LocalScalar with the same name. Also reusing the same declaration creates a fake dependency which might be confusing even if it might never be a problem for correctness.
Does it make sense?

Reply:
After slack chat: I misunderstood the problem. This PR is solving the issue of not being allowed to have LocalScalars with the same name in 2 different HorizontalExecutions. But this fix introduces the problem I was talking about: now if you have 2 distinct HorizontalExecutions with LocalScalars with the same name, that get merged into a single HorizontalExecution, you will still have a name clash.

Reply:
Yes. What I'm understanding is that we need a small change to `OnTheFlyMerging` in this PR that checks if there are `LocalScalars` with the same name, and if so renames one set of them to avoid the clash.

Reply:
Turns out after hours of investigation that it's not so easy... The pass very much assumes that the symbol table comes from the `symtable` keyword-argument from `visit_VerticalLoopSection`. Instead it needs to keep merging in the `horizontal_execution.symtable_` for each one it merges. The new `HorizontalExecution` needs to have a combination of symbols from that as well as the mapped scalars for any conflicted symbols. It's really rather tricky.

Reply:
What's still missing? Your test with the LocalScalar seems to pass?

Reply:
What I mentioned above still isn't implemented. It's creating one horizontal execution, but the code is wrong inside it.